devito/finite_differences/derivative.py

```diff
@@ -192,22 +195,42 @@ def T(self):
             adjoint = direct
 
         return Derivative(self.expr, *self.dims, deriv_order=self.deriv_order,
-                          fd_order=self.fd_order, side=self.side, stagger=self.stagger,
-                          transpose=adjoint)
+                          fd_order=self.fd_order, side=self.side, transpose=adjoint,
+                          x0=self.x0)
+
+    def eval_at(self, var):
+        """
+        Evaluates the derivative at the location of var. This is necessary for staggered
+        setup where one could have Eq(u(x + h_x/2), v(x).dx)) in which case v(x).dx
+        has to be computed at x=x + h_x/2.
+        """
+        x0 = {d1: d2 for d1, d2 in zip(var.dimensions, var.index_ref)}
+        return Derivative(self.expr, *self.dims, deriv_order=self.deriv_order,
+                          fd_order=self.fd_order, side=self.side,
+                          transpose=self.transpose, eval_at=self._eval_at, x0=x0)
 
     @property
     def evaluate(self):
-        expr = getattr(self.expr, 'evaluate', self.expr)
+        # If the expression is an addition, for example if expr was a derivative that
+        # was evaluated, split it and rebuild it as each term may have a different
+        # staggering and needs a separate FD computation
+        if self.expr.is_Add:
```

Comment:
this smells like a hack

Reply:
THis is not a hack in anyway and the comment explains it.

Reply:
I still think that checking whether is an Add is hacky, we would need a stronger, more reliable mechanism

would it work if were negated ? something like

-(... + ...)

now the root operation would be a Mul(..., -1)

Reply:
Fair point didn't think of that one. `Sparse` needs this kind of stuff too but may be heavy so maybe a TODO?

Reply:
I **think** we should deal with this, AFAIU this can be seriously dangerous